File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        left_max = height[0]
        right_max = height[len(height) - 1]
        l, r = 0, len(height) - 1 
        ans = 0
        while l<r:
            if left_max <= right_max:
                l+=1
                left_max = max(left_max, height[l])
                ans+=(left_max - height[l])
            else :
                r-=1
                right_max = max(right_max, height[r])
                ans+=(right_max - height[r])
        return ans

File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int l = 0, r = height.size()-1, lm = height[0], rm = height[r], ans = 0;
        while(l<r){
            if(lm<rm){
                l++;
                lm = max(lm, height[l]);
                ans+=lm - height[l];
            }
            else{
                r--;
                rm = max(rm, height[r]);
                ans+=rm - height[r];
            }
        }
        return ans;
    }
};